tower-http/src/cors.rs

```diff
@@ -0,0 +1,749 @@
+    }
+
+    /// Set the value of the [`Access-Control-Allow-Headers`][mdn] header.
+    ///
+    /// ```
+    /// use tower_http::cors::CorsLayer;
+    /// use http::header::HeaderValue;
+    ///
+    /// let layer = CorsLayer::new().allow_headers(vec![
+    ///     "*".parse().unwrap(),
+    /// ]);
+    /// ```
+    ///
+    /// By default the header will be set to `*`.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers
+    pub fn allow_headers<I>(mut self, headers: I) -> Self
+    where
+        I: IntoIterator<Item = HeaderValue>,
+    {
+        self.allow_headers = separated_by_commas(
+            headers
+                .into_iter()
+                .map(|value| value.to_str().unwrap().to_string()),
+        );
+        self
+    }
+
+    /// Set the value of the [`Access-Control-Max-Age`][mdn] header.
+    ///
+    /// ```
+    /// use tower_http::cors::CorsLayer;
+    /// use std::time::Duration;
+    ///
+    /// let layer = CorsLayer::new().max_age(Duration::from_secs(60) * 10);
+    /// ```
+    ///
+    /// By default the header will not be set will disables caching and will require a preflight
+    /// call for all requests.
+    ///
+    /// Note that each browser has a maximum internal value that takes precedence when the
+    /// Access-Control-Max-Age is greater. For more details see [mdn].
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age
+    pub fn max_age(mut self, max_age: Duration) -> Self {
+        self.max_age = Some(max_age.as_secs().into());
+        self
+    }
+
+    /// Set the value of the [`Access-Control-Allow-Methods`][mdn] header.
+    ///
+    /// ```
+    /// use tower_http::cors::CorsLayer;
+    /// use http::Method;
+    ///
+    /// let layer = CorsLayer::new().allow_methods(vec![Method::GET, Method::POST]);
+    /// ```
+    ///
+    /// All methods can be allowed with
+    ///
+    /// ```
+    /// use tower_http::cors::{CorsLayer, Any};
+    ///
+    /// let layer = CorsLayer::new().allow_methods(Any);
+    /// ```
+    ///
+    /// By default the header will be set to `GET, POST, OPTIONS`.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Methods
+    pub fn allow_methods<T>(mut self, methods: T) -> Self
+    where
+        T: Into<AnyOr<Vec<Method>>>,
+    {
+        self.allow_methods = match methods.into().0 {
+            AnyOrInner::Any => "*".parse().unwrap(),
+            AnyOrInner::Value(methods) => separated_by_commas(methods),
+        };
+        self
+    }
+
+    /// Set the value of the [`Access-Control-Allow-Origin`][mdn] header.
+    ///
+    /// ```
+    /// use tower_http::cors::{CorsLayer, Origin};
+    ///
+    /// let layer = CorsLayer::new().allow_origin(Origin::exact(
+    ///     "http://example.com".parse().unwrap(),
+    /// ));
+    /// ```
+    ///
+    /// Multiple origins can be allowed with
```

Comment:
Probably good to note that calling `.allow_origin` multiple times will not do the same thing.

Reply:
Actually the docs in general could be more explicit about a lot of things. I'll do another pass over them later.